src/Interpreters/HashJoin/KnownRowsHolder.h

```diff
@@ -15,7 +19,7 @@ template<>
 class KnownRowsHolder<true>
 {
 public:
-    using Type = std::pair<const Columns *, DB::RowRef::SizeT>;
+    using Type = PairNoInit<const Columns *, DB::RowRef::SizeT>;
```

Comment:
We create a new `KnownRowsHolder` for each row inside `joinRightColumns`. Each such object has a static array of 16 pairs, each pair is 16 bytes in size. As a result, we pay the cost equivalent to `memset(addr, 0, 256)` on each loop iteration to zero-initialize it. It is not good.